lib/queue.js

```diff
@@ -46,255 +41,242 @@ FetchQueue.prototype.add = function(protocol, domain, port, path, depth, callbac
-    callback = callback && callback instanceof Function ? callback : function() {};
-
+    callback = callback instanceof Function ? callback : noop;
     port = port !== 80 ? ":" + port : "";
 
     var url = (protocol + "://" + domain + port + path).toLowerCase();
 
     if (this.scanIndex[url]) {
         callback(null, 1);
-        return 1;
+    } else {
+        this.scanIndex[url] = true;
+        callback(null, 0);
     }
+};
 
-    this.scanIndex[url] = true;
-    callback(null, 0);
-    return 0;
-
+// Get item from queue
+FetchQueue.prototype.get = function(index, callback) {
+    callback = callback instanceof Function ? callback : noop;
+    callback(null, this[index]);
 };
 
 // Get last item in queue...
 FetchQueue.prototype.last = function(callback) {
-    callback = callback && callback instanceof Function ? callback : function() {};
-    var item,
-        self = this;
+    callback = callback instanceof Function ? callback : noop;
+    var queue = this;
 
-    item = self[self.length - 1];
-    callback(null, item);
-    return item;
-};
+    queue.getLength(function(error, length) {
+        if (error) {
+            return callback(error);
+        }
 
-// Get item from queue
-FetchQueue.prototype.get = function(id, callback) {
-    callback = callback && callback instanceof Function ? callback : function() {};
-    var item,
-        self = this;
-
-    if (!isNaN(id) && self.length > id) {
-        item = self[id];
-        callback(null, item);
-        return item;
-    }
+        queue.get(length - 1, function(error, item) {
+            if (error) {
+                callback(error);
+            } else {
+                callback(null, item);
+            }
+        });
+    });
 };
 
 // Get first unfetched item in the queue (and return its index)
 FetchQueue.prototype.oldestUnfetchedItem = function(callback) {
-    callback = callback && callback instanceof Function ? callback : function() {};
-    var item,
-        self = this;
-
-    for (var itemIndex = self.oldestUnfetchedIndex; itemIndex < self.length; itemIndex++) {
-        if (self[itemIndex].status === "queued") {
-            self.oldestUnfetchedIndex = itemIndex;
-            item = self[itemIndex];
-            callback(null, item);
-            return item;
+    callback = callback instanceof Function ? callback : noop;
+    var queue = this;
+
+    queue.getLength(function(error, length) {
+        function getItem (itemIndex) {
+            queue.get(itemIndex, function(error, item) {
+                if (error) {
+                    callback(error);
+                } else if (item.status === "queued") {
+                    queue.oldestUnfetchedIndex = itemIndex;
+                    callback(null, item);
+                } else if (itemIndex + 1 < length) {
+                    getItem(itemIndex + 1);
+                } else {
+                    callback(new Error("No unfetched items remain."));
```

Comment:
The fact that we call back with error here is a bit odd (and I'm wondering why I did it like that in the first place.)

Could we call back with (null, null)? Would that be understandable? Or is it truly an error condition?